configure script incorrectly assumes functions are available with some conditions

Working on the [CoreFX PAL fixes for AIX](https://github.com/dotnet/corefx/pull/30012), when I noticed that it was incorrectly assuming that sendfile was available, confusing CoreFX. Example, with a comparison against a case where it wasn't successful and thus caught by the linker:

```
configure:26246: checking for sendfile with 6 arguments
configure:26266: gcc -c -Wall -Wunused -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes  -Wmissing-prototypes -Wnested-externs -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -g -O2 -std=gnu99 -fno-strict-aliasing -fwrapv -DMONO_DLL_EXPORT -Wno-unused-but-set-variable -g -Wall -Wunused -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes  -Wmissing-prototypes -Wnested-externs -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -D__PASE__ -I/opt/mono/include -maix64 -DGC_AIX_THREADS -D_ALL_SOURCE -D_THREAD_SAFE -D_LARGE_FILES -D_REENTRANT -g  conftest.c >&5
conftest.c:251:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
 main ()
 ^~~~
conftest.c: In function 'main':
conftest.c:254:16: warning: implicit declaration of function 'sendfile' [-Wimplicit-function-declaration]
   int result = sendfile(0, 0, 0, NULL, NULL, 0);
                ^~~~~~~~
conftest.c:254:3: warning: nested extern declaration of 'sendfile' [-Wnested-externs]
   int result = sendfile(0, 0, 0, NULL, NULL, 0);
   ^~~
conftest.c:254:7: warning: unused variable 'result' [-Wunused-variable]
   int result = sendfile(0, 0, 0, NULL, NULL, 0);
       ^~~~~~
configure:26266: $? = 0
configure:26268: result: yes
configure:26282: checking for fcopyfile
configure:26282: gcc -o conftest -Wall -Wunused -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes  -Wmissing-prototypes -Wnested-externs -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -g -O2 -std=gnu99 -fno-strict-aliasing -fwrapv -DMONO_DLL_EXPORT -Wno-unused-but-set-variable -g -Wall -Wunused -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes  -Wmissing-prototypes -Wnested-externs -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -D__PASE__ -I/opt/mono/include -maix64 -DGC_AIX_THREADS -D_ALL_SOURCE -D_THREAD_SAFE -D_LARGE_FILES -D_REENTRANT -g  -maix64 conftest.c -lm   -lpthread >&5
conftest.c:268:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
 char fcopyfile ();
 ^~~~
conftest.c:277:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
 main ()
 ^~~~
ld: 0711-317 ERROR: Undefined symbol: .fcopyfile
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
configure:26282: $? = 1
configure: failed program was:
[...]
```

The difference between the two; one checks as a code fragment while the other lets autotools handle it. In source, on configure.ac+3072:

```
        AC_MSG_CHECKING(for sendfile with 6 arguments)
        AC_TRY_COMPILE([
                #include <stdlib.h>
                #include <sys/types.h>
                #include <sys/socket.h>
                #include <sys/uio.h>
        ], [
                int result = sendfile(0, 0, 0, NULL, NULL, 0);
        ],[
                AC_MSG_RESULT(yes)
                AC_DEFINE(HAVE_SENDFILE_6, 1, [sendfile with 6 arguments])
        ], [
                AC_MSG_RESULT(no)
        ])

        AC_CHECK_FUNC(fcopyfile,     [AC_DEFINE(HAVE_FCOPYFILE, 1, [fcopyfile])])
```

It seems that the linker catches it on the `AC_CHECK_FUNC` case, whereas the `AC_TRY_COMPILE` case merely makes an object without linking with the `-c` option passed to GCC. (FWIW, the "4 arguments" case for detecting sendfile for CoreFX fails because of a missing `sendfile.h` header - basically again, that case only handles compile errors.)

My guess for possible solutions are either:

1. Make `AC_TRY_COMPILE` try linking

2. Pass the compiler flag that makes GCC error out on implied definitions; though there *are* a few cases where fixincludes included with GCC has mangled system headers to not include something in headers that exists in libc, but I believe I caught them all already and explicitly make explicit declarations whenever GCC was being stupid

I'm not an autotools expert, so IDK how to implement either of these.

```
configure:26246: checking for sendfile with 6 arguments
configure:26266: gcc -o conftest -Wall -Wunused -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes  -Wmissing-prototypes -Wnested-externs -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -g -O2 -std=gnu99 -fno-strict-aliasing -fwrapv -DMONO_DLL_EXPORT -Wno-unused-but-set-variable -g -Wall -Wunused -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes  -Wmissing-prototypes -Wnested-externs -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Wno-format-zero-length -D__PASE__ -I/opt/mono/include -maix64 -DGC_AIX_THREADS -D_ALL_SOURCE -D_THREAD_SAFE -D_LARGE_FILES -D_REENTRANT -g  -maix64 conftest.c -lm   -lpthread >&5
conftest.c:251:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
 main ()
 ^~~~
conftest.c: In function 'main':
conftest.c:254:16: warning: implicit declaration of function 'sendfile' [-Wimplicit-function-declaration]
   int result = sendfile(0, 0, 0, NULL, NULL, 0);
                ^~~~~~~~
conftest.c:254:3: warning: nested extern declaration of 'sendfile' [-Wnested-externs]
   int result = sendfile(0, 0, 0, NULL, NULL, 0);
   ^~~
conftest.c:254:7: warning: unused variable 'result' [-Wunused-variable]
   int result = sendfile(0, 0, 0, NULL, NULL, 0);
       ^~~~~~
ld: 0711-317 ERROR: Undefined symbol: .sendfile
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
configure:26266: $? = 1
configure: failed program was:
```

So far so good; I'll wait until it reaches CoreFX PAL before I `^C` and branch.

Address of the thing is still a warning:

```$ cat 1.c
#include <string.h>

int (*p)(const char*) = &strlen;jaymac1:mkbundle jay$ cc -c 1.c
1.c:3:7: warning: incompatible function pointer types initializing 'int (*)(const char *)' with an expression of type 'unsigned long (*)(const char *)' [-Wincompatible-function-pointer-types]
int (*p)(const char*) = &strlen;
      ^                 ~~~~~~~
1 warning generated.
```
So suggest the compile as C++ approach, as demonstrated elsewhere.

--- a/configure.ac
+++ b/configure.ac
@@ -3070,7 +3070,7 @@ if test x$host_win32 = xno; then
 	])
 
 	AC_MSG_CHECKING(for sendfile with 6 arguments)
-	AC_TRY_COMPILE([
+	AC_TRY_LINK([
 		#include <stdlib.h>
 		#include <sys/types.h>
 		#include <sys/socket.h>

